### ISSUE TITLE
Variables can be used before they are declared

### ISSUE THREAD

**Issue by redsquirrel** [ethereum/solidity#3192]
I was surprised this compiled:

```
pragma solidity 0.4.18;

contract Sandbox {
    function foo() public pure returns(uint) {
        uint baz = bar + bar;
        uint bar = 10;
        return baz;
    }
}
```

**Comment by roadriverrail** [comment]
This has been a part of the language for a while.  Actually, a few features
I've written have had to take it into account.  Despite it being pretty
unappealing (to me and you), I'm not sure it can be classified as an issue.

**Comment by redsquirrel** [comment]
What are the reasons to support the ability to use a variable before it's declared?

**Comment by o0ragman0o** [comment]
Declarations are hoisted in a similar way that JS hoists.

**Comment by axic** [comment]
This is being addressed in #1679. I'm closing it as it is part of the language taking its basic scoping rules from Javascript. I'm sure the documentation explains this, but if not please reopen the issue.
